### DISCUSSION TITLE
[QUESTION] How to judge the collection of local candidates is finished?

### DISCUSSION THREAD

**Discussion by echomlv** [awslabs/amazon-kinesis-video-streams-webrtc-sdk-c#1732] [Q&A]
In our application scenario, we want to send the answer after collecting the candidates，and we found the following callback function，it says  "if PCHAR is NULL then candidate gathering is done." .

```c
/**
 * @brief RtcOnIceCandidate is fired when new iceCandidate is found. if PCHAR is NULL then candidate gathering is done.
 *
 * Reference: https://www.w3.org/TR/webrtc/#dom-rtcpeerconnection-onicecandidate
 */
typedef VOID (*RtcOnIceCandidate)(UINT64, PCHAR);
```

But, after testing, it seems that it is related to the timeout period, and NULL will only be returned when it times out.
guessed by modifying the value of `iceLocalCandidateGatheringTimeout` configuration and printing logs

```c
    UINT32 iceLocalCandidateGatheringTimeout; //!< Maximum time ice will wait for gathering STUN and RELAY candidates. Once
                                              //!< it's reached, ice will proceed with whatever candidate it current has. Use default value if 0.
```

Is there any way to judge like google's webrtc implementation ？
```javascript
 /* --method 1*/
let pc = new RTCPeerConnection();
pc.onicegatheringstatechange = ev => {
  let connection = ev.target;

  switch(connection.iceGatheringState) {
    case "gathering":
      /* collection of candidates has begun */
      break;
    case "complete":
      /* collection of candidates is finished */
      break;
  }
}


 /* --method 2*/
let pc = new RTCPeerConnection();
pc.onicecandidate = function(event) {
  if (event.candidate) {
    /* Code for each candidate */
    /* Send the candidate to the remote peer */
  } else {
    /* All ICE candidates have been sent */
  }
}

```

Need help on this.

Thanks.

**Comment by alitofresko** [comment]
It is a very interesting stuff.
I too am in the same situation.
I'm sure it's not a default behaviour, but for some reason we got a gathering complete callback only after the timeout.
I hope someone can show us the correct way to solve this problem.
